### PR TITLE
Switch Implementation

### LINKED ISSUE #39: Moving to an alternative implementation
I have gotten the permission from @camio  to adopt the [reference implementation](https://godbolt.org/z/5P78aG5xE) by @gnzlbg in the paper.

I propose we set-a-side current implementation as there's significant work for fixing and catching up.

Reference implementation will start to be synced in [ref-impl](https://github.com/bemanproject/inplace_vector/tree/ref-impl) branch. I will open a PR once it is reasonably integrated with current system.

Meanwhile, I suggest we halt all progress on PRs related to `include/.../inplace_vector.hpp` in the main branch.

Feel free (and please) help me by writing PRs against the ref_impl branch.

Task list:

Main implementation:
- [x] sync implementation to ref-impl
- [x] update cmake script if needed
- [x] lint implementation
- [x] update name space to beman

Test suite:
- [x] sync tests (or subset of test) to ref-impl
- [x] pass all tests

Documentation:
- [x] Determine new minimum C++ standard

@camio @DeveloperPaul123 @ClausKlein @jbab

### PR DESCRIPTION
**Pull request by wusatosi**
This PR switches current implementation to reference implementation in the paper by @gnzlbg .

Original link: https://godbolt.org/z/5P78aG5xE

Permission to switch this implementation was granted by @camio .

Fixed CI and included test suite from original implementation.

Let's not scrutinize the details of the implementation yet.

CC: @jbab

Fix #39

### CODE REVIEW

File: .github/workflows/ci_tests.yml [hunk 1]
```diff
@@ -31,13 +31,13 @@ jobs:
...
       fail-fast: false
       matrix:
-        platform: [ubuntu-latest]
+        platform: [ubuntu-24.04]
         compiler:
           - cpp: g++
             c: gcc
           - cpp: clang++
             c: clang
-        cpp_version: [17, 20, 23, 26]
```
Comment (DeveloperPaul123) [inline .github/workflows/ci_tests.yml:40]: Can we not keep this with the new ref impl? It requires C++23?

**Review by DeveloperPaul123 — APPROVED** [review]
Seems fine to me. I don't have the time to fully review the ref impl, but I assume it's fine.

**Comment by wusatosi** [comment]
> Seems fine to me. I don't have the time to fully review the ref impl, but I assume it's fine.

We will do that later. This is just switching to reference implementation

File: README.md [hunk 2]
```diff
@@ -56,7 +56,7 @@ template <int Capacity> inplace_vector<int, Capacity> fibonacci_to(int num) {
 
 ### Compiler support
 
-Building this repository requires **C++17** or later.
+Building this repository requires **C++23** or later.
```
Comment (ClausKlein) [inline README.md:59]: change below, line 73, too

**Comment by ClausKlein** [comment]
# Try this too!
```bash
cmake -S . -B build -G Ninja -DCMAKE_CXX_STANDARD=23 -DCMAKE_EXPORT_COMPILE_COMMANDS=ON --fresh
# ...
run-clang-tidy -p build -checks='-*,cppcoreguidelines-pro-*'
Enabled checks:
    cppcoreguidelines-pro-bounds-array-to-pointer-decay
    cppcoreguidelines-pro-bounds-constant-array-index
    cppcoreguidelines-pro-bounds-pointer-arithmetic
    cppcoreguidelines-pro-type-const-cast
    cppcoreguidelines-pro-type-cstyle-cast
    cppcoreguidelines-pro-type-member-init
    cppcoreguidelines-pro-type-reinterpret-cast
    cppcoreguidelines-pro-type-static-cast-downcast
    cppcoreguidelines-pro-type-union-access
    cppcoreguidelines-pro-type-vararg

...
```

File: .github/workflows/ci_tests.yml (hunk 1 above)
Comment (jbab) [reply .github/workflows/ci_tests.yml:40]: [from_range](https://en.cppreference.com/w/cpp/ranges/from_range), e.g., is C++23.
I'd suggest bringing the ref implementation to main first and then explore if and how earlier standards can be enabled.

File: README.md (hunk 2 above)
Comment (wusatosi) [reply README.md:59]: Agh, missed this.

**Comment by wusatosi** [comment]
Hey @ClausKlein sorry I didn't reply to your message before merging.

This is simply to port the reference implementation. We might incorporate suggestions from clang-tidy but the main concern currently is to review the reference implementation against the paper.

I don't really think we should do clang-tidy check for test files, especially when this clearly checks exception throwing:
```
[1/3][2.9s] /usr/local/opt/llvm/bin/clang-tidy -checks=-*,bugprone-* -p=build /Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/inplace_vector.test.cpp
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/inplace_vector.test.cpp:75:7: warning: empty catch statements hide issues; to handle exceptions appropriately, consider re-throwing, handling, or avoiding catch altogether [bugprone-empty-catch]
   75 |     } catch (const std::out_of_range &) {
      |       ^
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/inplace_vector.test.cpp:83:7: warning: empty catch statements hide issues; to handle exceptions appropriately, consider re-throwing, handling, or avoiding catch altogether [bugprone-empty-catch]
   83 |     } catch (const std::out_of_range &) {
      |       ^
27827 warnings generated.
Suppressed 27825 warnings (27825 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
```

I don't think this is valuable for testing structs, also there's no instance of misuse reported.
```
[3/3][4.7s] /usr/local/opt/llvm/bin/clang-tidy -checks=-*,bugprone-* -p=build /Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:139:16: warning: 2 adjacent parameters of 'non_copyable' of convertible types are easily swapped by mistake [bugprone-easily-swappable-parameters]
  139 |   non_copyable(int i, double d) : i_(i), d_(d) {}
      |                ^~~~~~~~~~~~~~~
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:139:20: note: the first parameter in the range is 'i'
  139 |   non_copyable(int i, double d) : i_(i), d_(d) {}
      |                    ^
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:139:30: note: the last parameter in the range is 'd'
  139 |   non_copyable(int i, double d) : i_(i), d_(d) {}
      |                              ^
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:139:23: note: 'int' and 'double' may be implicitly converted
  139 |   non_copyable(int i, double d) : i_(i), d_(d) {}
      |                       ^
```

Why not throw exception for tests at main? I don't think try-catch the main function and return -1 is of any value for tests.
```
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:336:5: warning: an exception may be thrown in function 'main' which should not throw exceptions [bugprone-exception-escape]
  336 | int main() {
      |     ^
```

Reference implementation aimed at std-style implementation (originally under std namespace). We would like to move away from reserved identifier but it's not top priority.
```
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:338:42: warning: declaration uses identifier '__non_trivial', which is a reserved identifier [bugprone-reserved-identifier]
  338 |     using beman::__iv_detail::__storage::__non_trivial;
      |                                          ^~~~~~~~~~~~~
      |                                          _non_trivial
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:339:42: warning: declaration uses identifier '__trivial', which is a reserved identifier [bugprone-reserved-identifier]
  339 |     using beman::__iv_detail::__storage::__trivial;
      |                                          ^~~~~~~~~
      |                                          _trivial
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:340:42: warning: declaration uses identifier '__zero_sized', which is a reserved identifier [bugprone-reserved-identifier]
  340 |     using beman::__iv_detail::__storage::__zero_sized;
      |                                          ^~~~~~~~~~~~
      |                                          _zero_sized
```

The intent is literally to test the move-from object:
```
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:486:11: warning: 'a' used after it was moved [bugprone-use-after-move]
  486 |     CHECK(a.size() == std::size_t{3});
      |           ^
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:484:7: note: move occurred here
  484 |     b = std::move(a);
      |       ^
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:495:11: warning: 'a' used after it was moved [bugprone-use-after-move]
  495 |     CHECK(a.size() == std::size_t{3});
      |           ^
/Users/clausklein/Workspace/cpp/beman-project/inplace_vector/tests/beman/inplace_vector/ref_impl.test.cpp:493:25: note: move occurred here
  493 |     vector<MoveOnly, 3> b(std::move(a));
      |                         ^
```

I wish this line by line explanation shows why I am not a big advocate for clang-tidy, I am sorry but this report is not helpful at all to our implementation.

**Pull request merged**